### PR TITLE
fix: optional service account

### PR DESCRIPTION
**Pull request by rasheedamir**
don't add service account until user asks for it! it will get default service account
Fixes https://github.com/stakater/application/issues/195

### CODE REVIEW

**Comment by stakater-user** [comment]
@rasheedamir Validation Successful

**Comment by rasheedamir** [comment]
@AsfaMumtaz plz update the changelog as well

File: application/values-test.yaml
Comment (rasheedamir) [inline application/values-test.yaml:389]: this shouldn't be hardcoded instead generated dynamically

File: application/templates/deployment.yaml
Comment (d3adb5) [inline application/templates/deployment.yaml:281]: I think we can leverage order of evaluation here to simplify this line. This seems to be working as of Helm v3.11.1:

```suggestion
      {{- if and .Values.rbac .Values.rbac.serviceAccount .Values.rbac.serviceAccount.name }}
```

However I think we should instead just make sure `.Values.rbac.serviceAccount` is at least `{}` in our `values.yaml`, **which we already do!** So I recommend going simply with:

```suggestion
      {{- with .Values.rbac.serviceAccount.name }}
```

And using `.` within. This is because to break the above line, one would have to provide overrides like:

```yaml
rbac:
```

or

```yaml
rbac:
  serviceAccount:
```

At that point, I'd expect the chart user to know what they're doing and understand what to do when template rendering fails.

File: application/values-test.yaml
Comment (d3adb5) [reply application/values-test.yaml:389]: These values are used to ensure the templates are rendered properly during a `helm install`, right? What's your idea for generating this value dynamically? :thinking:

**Review by d3adb5 — CHANGES_REQUESTED** [review]

**Comment by unixfox** [comment]
Looking forward for this PR because currently it's not possible to use this chart if you don't want to have any RBAC involved.

File: application/templates/deployment.yaml
Comment (mustafaStakater) [reply application/templates/deployment.yaml:281]: We need to define default service account `{{ template "application.name" $ }}` in else block incase `.Values.rbac.serviceAccount.name` isnt defined, using `if` is fine

Comment (mustafaStakater) [reply application/templates/deployment.yaml:281]: What do you think about going forward as below

```suggestion
      {{- if .Values.rbac.serviceAccount.enabled }}
      {{- if $.Values.rbac.serviceAccount.name }}
      serviceAccountName: {{ $.Values.rbac.serviceAccount.name }}
      {{- else }}
      serviceAccountName: {{ template "application.name" $ }}
      {{- end }}
      {{- end }}
```

**Review by aslafy-z — APPROVED** [review]

File: application/templates/deployment.yaml
Comment (d3adb5) [inline application/templates/deployment.yaml:300]: I think it's worth adding a chart unit test to check that the `serviceAccountName` field is empty. We have three possible branches here: populated and proxied from values, populated via the fallback (`template "application.name" $`), and empty.

At the very least we should test that it's empty when disabled:

```yaml
- it: yields empty service account name when disabled
  set:
    rbac.serviceAccount.enabled: false
  asserts:
    - isEmpty:
        path: spec.template.spec.serviceAccountName
```

**Review by d3adb5 — CHANGES_REQUESTED** [review]

File: application/templates/deployment.yaml
Comment (aslafy-z) [reply application/templates/deployment.yaml:297]: I added tests to cover the three branches. Let me know if it fits you :)

**Comment by rasheedamir** [comment]
@d3adb5 is it good to merge now?

**Comment by aslafy-z** [comment]
@rasheedamir I would like to wait for @d3adb5 approval first, but yes, I think we're good to go.

**Review by d3adb5 — APPROVED** [review]
LGTM!

**Pull request merged**
